# III. SQL-анализ и продуктовые метрики

**Цель:** провести углублённый SQL-анализ очищенных данных, рассчитать ключевые продуктовые и финансовые метрики, сформировать агрегированные таблицы для проверки гипотез и бизнес-выводов.

---
## 1. Импорт библиотек и подключение к базе данных
---

In [1]:
import pandas as pd
import os

from dotenv import load_dotenv
from urllib.parse import quote_plus

# Импорт модуля для создания соединения с базой
from sqlalchemy import create_engine

In [2]:
# Настройка отображения SQL-результатов в виде таблицы
import prettytable
prettytable.DEFAULT = prettytable.TableStyle

Для выполнения SQL-запросов прямо в ячейках ноутбука (без python) необходимо загрузить расширение ipython-sql.
Оно добавляет команду **%sql** — для однострочных запросов и команду **%%sql** — для многострочных SQL-запросов.

In [3]:
# Загрузим расширение ipython-sql для выполнения SQL-запросов прямо в ноутбуке
%load_ext sql

Подключимся к PostgreSQL из ноутбука:

In [4]:
load_dotenv()

user = os.getenv("DB_USER")
password = quote_plus(os.getenv("DB_PASS"))
host = os.getenv("DB_HOST", "localhost")
port = int(os.getenv("DB_PORT", 5432))
database = os.getenv("DB_NAME")

# Создадим соединение
engine = create_engine(f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}")
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"

# Проверим соединение
with engine.connect() as conn:
    print("Успешное подключение к PostgreSQL")

Успешное подключение к PostgreSQL


In [ ]:
data_path = "../data/cleaned/kicksharing_clean.csv"

df = pd.read_csv(data_path)

df.to_sql("rides", engine, index=False, if_exists="replace")

print("Таблица rides успешно загружена в PostgreSQL")

In [6]:
# Подключимся к созданной базе данных через ipython-sql
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"
%sql $conn_str

Проверим работу SQL-запроса с помощью "магической" команды %sql:

In [7]:
%sql SELECT * FROM rides LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
5 rows affected.


[(252614779, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.805660774955616, '2024-04-17 16:45:16', '2024-04-17 16:45:20.641980', '2024-04-17 16:51:26.403392', 3, '2024-04-17 16:45:20.641980', 97.94, 0.0, 6.096023533333334, 7.929701424709663, 0.0, 16, 2, 4),
 (252613066, 297828357, 'M', 31, 'Unknown', 'UNM', 'ХАНТЫ-МАНСИЙСКИЙ АВТОНОМНЫЙ ОКРУГ - ЮГРА АО', 7.99, 50.0, 300.0, 'SL', 0.237860583289626, '2024-04-22 15:46:04', '2024-04-22 15:46:10.062458', '2024-04-22 15:55:04.011996', 3, '2024-04-22 15:46:10.062458', 121.91, 1.0, 8.899158966666667, 1.6037060413050748, 0.0082027725371175, 15, 0, 4),
 (252621809, 145875613, 'M', 35, 'GRD', 'DIV', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.282096271526722, '2024-04-22 15:46:53', '2024-04-22 15:47:00.283797', '2024-04-22 15:56:28.643724', 3, '2024-04-22 15:47:00.283797', 129.9, 1.0, 9.47266545, 1.7868018648967827, 0.0076982294072363, 15, 0, 4),
 (252632392, 328153536, 'M', 25, 'Unknown', 'Unknown', 'Г САНКТ-ПЕТЕРБУРГ', 6.99, 30.0, 300.0, 'SL', 0.303165913380335, '2024-04-23 06:53:41', '2024-04-23 06:53:45.475088', '2024-04-23 06:56:59.051051', 3, '2024-04-23 06:53:45.475088', 57.96, 0.0, 3.22626605, 5.638082700222475, 0.0, 6, 1, 4),
 (252633271, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.49, 50.0, 300.0, 'SL', 0.822775782880499, '2024-04-23 09:46:17', '2024-04-23 09:46:22.008459', '2024-04-23 09:50:38.041837', 3, '2024-04-23 09:46:22.008459', 87.45, 0.0, 4.267222966666667, 11.568776077194888, 0.0, 9, 1, 4)]

---
## 2. Базовые метрики
---

In [8]:
%%sql
SELECT
    COUNT(*) as total_orders,
    COUNT(DISTINCT(party_rk)) as users,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback_rate
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(355637, 58033, Decimal('2.478'), Decimal('129.82'), Decimal('11.65'), Decimal('0.23'))]

- Общее количество поездок: 355637.
- Количество уникальных пользователей: 58033.
- Средний километраж поездки: 2,478.
- Средняя стоимость поездки (в рублях): 129,82.
- Средняя длительность поездки (в минутах): 11,65.
- Средняя доля кэшбэка от стоимости поездки: 0,23.

---
## 3. Финансовые метрики
---

Посчитаем следующие финансовые метрики:
- выручку сервиса;
- сумму выплаченного кэшбэка;
- долю выплаченного кэшбэка от выручки;
- средний доход с пользователя.

In [9]:
%%sql
SELECT
    ROUND(SUM(nominal_price_rub_amt)::NUMERIC, 2) as total_revenue,
    ROUND(SUM(loyalty_accrual_rub_amt)::NUMERIC, 2) as total_cashback,
    ROUND((SUM(loyalty_accrual_rub_amt) * 1.0 / SUM(nominal_price_rub_amt))::NUMERIC, 2) as cashback_share,
    ROUND(SUM(nominal_price_rub_amt)::NUMERIC * 1.0 / COUNT(DISTINCT party_rk), 2) as arpu
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('46169712.86'), Decimal('11435784.20'), Decimal('0.25'), Decimal('795.58'))]

---
## 4. Группировки и агрегаты
---

По полу:

In [10]:
%%sql
SELECT
    gender_cd,
    COUNT(DISTINCT party_rk) as users,
    COUNT(*) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_duration,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback
FROM rides
GROUP BY gender_cd
ORDER BY 2 DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('M', 46835, 300822, Decimal('2.479'), Decimal('127.91'), Decimal('11.43'), Decimal('0.23')),
 ('F', 10762, 52194, Decimal('2.500'), Decimal('141.28'), Decimal('13.03'), Decimal('0.25')),
 ('U', 436, 2621, Decimal('1.884'), Decimal('120.83'), Decimal('10.14'), Decimal('0.23'))]

По регионам (так как регионов проживания больше 200, выведем топ-10 регионов по количеству поездок):

In [11]:
%%sql
SELECT
    lvn_state_nm as region,
    COUNT(*) as total_rides,
    COUNT(DISTINCT party_rk) as users,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback
FROM rides
WHERE lvn_state_nm <> 'Unknown'
GROUP BY lvn_state_nm
ORDER BY 2 DESC
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[('Г МОСКВА', 82120, 10833, Decimal('2.096'), Decimal('132.65'), Decimal('9.99'), Decimal('0.22')),
 ('МОСКОВСКАЯ ОБЛ', 35983, 5483, Decimal('2.214'), Decimal('132.80'), Decimal('10.70'), Decimal('0.23')),
 ('Г САНКТ-ПЕТЕРБУРГ', 34298, 5079, Decimal('2.664'), Decimal('128.11'), Decimal('10.87'), Decimal('0.23')),
 ('СВЕРДЛОВСКАЯ ОБЛ', 26128, 3001, Decimal('2.351'), Decimal('106.50'), Decimal('11.48'), Decimal('0.22')),
 ('КРАСНОДАРСКИЙ КРАЙ', 14612, 2013, Decimal('3.041'), Decimal('129.88'), Decimal('12.75'), Decimal('0.21')),
 ('РЕСП ТАТАРСТАН', 8703, 1617, Decimal('2.699'), Decimal('116.25'), Decimal('12.98'), Decimal('0.24')),
 ('РЕСП БАШКОРТОСТАН', 7564, 1442, Decimal('2.336'), Decimal('125.42'), Decimal('12.95'), Decimal('0.22')),
 ('НИЖЕГОРОДСКАЯ ОБЛ', 6777, 1324, Decimal('2.386'), Decimal('130.89'), Decimal('12.38'), Decimal('0.22')),
 ('НОВОСИБИРСКАЯ ОБЛ', 6146, 1046, Decimal('2.185'), Decimal('103.74'), Decimal('11.84'), Decimal('0.24')),
 ('ЛЕНИНГРАДСКАЯ ОБЛ', 5057, 860, Decimal('2.562'), Decimal('125.19'), Decimal('10.69'), Decimal('0.22'))]

По моделям самокатов:

In [12]:
%%sql
SELECT
    transport_model,
    COUNT(*) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback
FROM rides
GROUP BY transport_model
ORDER BY 2 DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
14 rows affected.


[('SL', 219011, Decimal('2.487'), Decimal('131.69'), Decimal('11.01'), Decimal('0.23')),
 ('Max', 61752, Decimal('2.695'), Decimal('133.97'), Decimal('13.12'), Decimal('0.23')),
 ('E', 59967, Decimal('2.246'), Decimal('114.28'), Decimal('11.87'), Decimal('0.23')),
 ('Urban E-Bike', 3932, Decimal('2.662'), Decimal('132.69'), Decimal('11.23'), Decimal('0.26')),
 ('ES400', 3668, Decimal('2.441'), Decimal('142.12'), Decimal('14.47'), Decimal('0.26')),
 ('ES200', 2866, Decimal('2.065'), Decimal('147.91'), Decimal('13.78'), Decimal('0.24')),
 ('ES100', 2398, Decimal('2.151'), Decimal('158.90'), Decimal('16.47'), Decimal('0.27')),
 ('MK', 929, Decimal('2.453'), Decimal('155.65'), Decimal('16.15'), Decimal('0.21')),
 ('ES-08S', 730, Decimal('2.732'), Decimal('202.08'), Decimal('19.51'), Decimal('0.23')),
 ('Techno', 196, Decimal('1.872'), Decimal('144.83'), Decimal('12.80'), Decimal('0.25')),
 ('ES', 92, Decimal('2.357'), Decimal('178.33'), Decimal('19.71'), Decimal('0.29')),
 ('LB', 70, Decimal('2.864'), Decimal('160.48'), Decimal('16.64'), Decimal('0.23')),
 ('B1D', 24, Decimal('2.567'), Decimal('273.31'), Decimal('24.83'), Decimal('0.39')),
 ('Lucky', 2, Decimal('0.293'), Decimal('161.25'), Decimal('61.96'), Decimal('0.55'))]

---
## 5. Временные и поведенческие метрики
---

Количество поездок по месяцам:

In [13]:
%%sql
SELECT
    month as num_month,
    CASE
        WHEN month = 4 THEN 'April'
        WHEN month = 5 THEN 'May'
        WHEN month = 6 THEN 'June'
        WHEN month = 7 THEN 'July'
        WHEN month = 8 THEN 'August'
        WHEN month = 9 THEN 'September'
        WHEN month = 10 THEN 'October'
    END as month_name,
    COUNT(*) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_duration,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback,
    ROUND(SUM(nominal_price_rub_amt)::NUMERIC, 2) as revenue
FROM rides
GROUP BY month, month_name
ORDER BY revenue DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
7 rows affected.


[(9, 'September', 100163, Decimal('2.751'), Decimal('132.70'), Decimal('11.30'), Decimal('0.27'), Decimal('13291911.15')),
 (8, 'August', 97284, Decimal('2.738'), Decimal('132.41'), Decimal('12.12'), Decimal('0.26'), Decimal('12880895.04')),
 (7, 'July', 73564, Decimal('2.330'), Decimal('127.48'), Decimal('11.96'), Decimal('0.17'), Decimal('9378147.87')),
 (10, 'October', 43041, Decimal('1.843'), Decimal('123.92'), Decimal('9.82'), Decimal('0.28'), Decimal('5333672.10')),
 (6, 'June', 37215, Decimal('2.131'), Decimal('125.16'), Decimal('12.65'), Decimal('0.13'), Decimal('4657738.63')),
 (5, 'May', 4324, Decimal('2.179'), Decimal('143.93'), Decimal('13.61'), Decimal('0.11'), Decimal('622335.89')),
 (4, 'April', 46, Decimal('1.400'), Decimal('108.96'), Decimal('9.09'), Decimal('0.07'), Decimal('5012.18'))]

Самые прибыльные месяцы: сентябрь и август.

Количество поездок по дням недели:

In [14]:
%%sql
SELECT
    TO_CHAR(book_start_dttm::TIMESTAMP, 'Day') AS day_name,
    COUNT(*) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_duration,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback,
    ROUND(COUNT(*) * AVG(nominal_price_rub_amt)::NUMERIC, 2) as revenue
FROM rides
GROUP BY day_name
ORDER BY total_rides DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
7 rows affected.


[('Friday   ', 56244, Decimal('2.478'), Decimal('127.75'), Decimal('11.47'), Decimal('0.23'), Decimal('7185245.59')),
 ('Thursday ', 53829, Decimal('2.395'), Decimal('124.81'), Decimal('11.09'), Decimal('0.23'), Decimal('6718377.84')),
 ('Wednesday', 53274, Decimal('2.409'), Decimal('126.06'), Decimal('11.21'), Decimal('0.23'), Decimal('6715963.85')),
 ('Tuesday  ', 51498, Decimal('2.453'), Decimal('125.83'), Decimal('11.24'), Decimal('0.23'), Decimal('6480248.69')),
 ('Monday   ', 49413, Decimal('2.414'), Decimal('124.12'), Decimal('11.09'), Decimal('0.23'), Decimal('6133012.16')),
 ('Saturday ', 48505, Decimal('2.616'), Decimal('141.53'), Decimal('12.87'), Decimal('0.24'), Decimal('6864799.03')),
 ('Sunday   ', 42874, Decimal('2.618'), Decimal('141.63'), Decimal('12.90'), Decimal('0.24'), Decimal('6072065.70'))]

Пятница приносит максимальную выручку (7,18 млн руб.) за счёт большего числа поездок, несмотря на меньшую среднюю стоимость по сравнению с выходными.

Распределение по часам:

In [15]:
%%sql
SELECT
    hour,
    COUNT(order_rk) as rides_in_hour,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_duration,
    ROUND(AVG(cashback_rate)::NUMERIC, 2) as avg_cashback
FROM rides
GROUP BY hour
ORDER BY 2 DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
24 rows affected.


[(15, 31306, Decimal('2.620'), Decimal('138.32'), Decimal('12.20'), Decimal('0.23')),
 (16, 29005, Decimal('2.515'), Decimal('137.84'), Decimal('12.10'), Decimal('0.23')),
 (14, 26000, Decimal('2.599'), Decimal('136.97'), Decimal('12.30'), Decimal('0.23')),
 (17, 24023, Decimal('2.543'), Decimal('137.63'), Decimal('12.27'), Decimal('0.23')),
 (5, 23028, Decimal('2.245'), Decimal('107.72'), Decimal('9.09'), Decimal('0.22')),
 (18, 20346, Decimal('2.553'), Decimal('137.32'), Decimal('12.40'), Decimal('0.23')),
 (13, 19872, Decimal('2.565'), Decimal('139.35'), Decimal('12.52'), Decimal('0.24')),
 (6, 18405, Decimal('2.201'), Decimal('110.44'), Decimal('9.37'), Decimal('0.22')),
 (12, 17168, Decimal('2.546'), Decimal('133.56'), Decimal('12.55'), Decimal('0.24')),
 (19, 16886, Decimal('2.548'), Decimal('135.61'), Decimal('12.33'), Decimal('0.23')),
 (11, 16464, Decimal('2.469'), Decimal('131.20'), Decimal('12.21'), Decimal('0.24')),
 (10, 16208, Decimal('2.447'), Decimal('129.54'), Decimal('12.07'), Decimal('0.23')),
 (9, 14979, Decimal('2.394'), Decimal('125.33'), Decimal('11.66'), Decimal('0.23')),
 (4, 14277, Decimal('2.197'), Decimal('104.26'), Decimal('8.70'), Decimal('0.21')),
 (7, 14144, Decimal('2.222'), Decimal('114.84'), Decimal('10.14'), Decimal('0.22')),
 (8, 13328, Decimal('2.348'), Decimal('122.25'), Decimal('11.24'), Decimal('0.23')),
 (20, 12525, Decimal('2.621'), Decimal('136.33'), Decimal('12.61'), Decimal('0.23')),
 (21, 8681, Decimal('2.687'), Decimal('138.17'), Decimal('12.87'), Decimal('0.23')),
 (3, 5248, Decimal('2.112'), Decimal('107.65'), Decimal('8.38'), Decimal('0.21')),
 (22, 5029, Decimal('2.869'), Decimal('146.96'), Decimal('14.25'), Decimal('0.23')),
 (23, 2994, Decimal('2.900'), Decimal('150.11'), Decimal('14.64'), Decimal('0.24')),
 (2, 2086, Decimal('2.378'), Decimal('119.23'), Decimal('10.15'), Decimal('0.22')),
 (0, 2027, Decimal('2.872'), Decimal('148.03'), Decimal('14.55'), Decimal('0.24')),
 (1, 1608, Decimal('2.753'), Decimal('140.72'), Decimal('13.47'), Decimal('0.23'))]

- Самое активное время — вечернее (с 14 до 18), о чём было упомянуто ранее в EDA.
- В утреннее время (с 5 до 7 утра) поездки по километражу и стоимости меньше, чем дневные и вечерние.

---
## 6. Использование оконных функций и подзапросов
---

Топ-10 пользователей по выручке:

In [16]:
%%sql
SELECT
    party_rk as user_id,
    lvn_state_nm,
    COUNT(order_rk) as count_rides,
    ROUND(SUM(nominal_price_rub_amt)::NUMERIC, 2) as total_revenue,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    RANK() OVER(ORDER BY SUM(nominal_price_rub_amt) DESC) as rank
FROM rides
GROUP BY party_rk, lvn_state_nm
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(386134592, 'СВЕРДЛОВСКАЯ ОБЛ', 336, Decimal('54235.56'), Decimal('161.42'), 1),
 (759042976, 'РЕСП УДМУРТСКАЯ', 269, Decimal('47541.41'), Decimal('176.73'), 2),
 (910211081, 'РЕСП БАШКОРТОСТАН', 333, Decimal('47343.08'), Decimal('142.17'), 3),
 (284170812, 'Г САНКТ-ПЕТЕРБУРГ', 191, Decimal('36849.82'), Decimal('192.93'), 4),
 (636557947, 'СВЕРДЛОВСКАЯ ОБЛ', 300, Decimal('34166.70'), Decimal('113.89'), 5),
 (231464761, 'Г МОСКВА', 285, Decimal('33580.57'), Decimal('117.83'), 6),
 (666474893, 'МОСКОВСКАЯ ОБЛ', 187, Decimal('32211.73'), Decimal('172.26'), 7),
 (528704373, 'МОСКОВСКАЯ ОБЛ', 155, Decimal('30340.27'), Decimal('195.74'), 8),
 (391309281, 'Г САНКТ-ПЕТЕРБУРГ', 299, Decimal('29861.78'), Decimal('99.87'), 9),
 (976560193, 'Г МОСКВА', 166, Decimal('29747.06'), Decimal('179.20'), 10)]

Активные пользователи приносят существенную часть выручки — стоит рассмотреть программы лояльности или таргет-акции для них.

Топ-10 регионов по выручке:

In [17]:
%%sql
SELECT
    lvn_state_nm,
    COUNT(order_rk) as total_rides,
    ROUND(SUM(nominal_price_rub_amt)::NUMERIC, 2) as total_revenue,
    RANK() OVER(ORDER BY SUM(nominal_price_rub_amt)::NUMERIC DESC) as rank
FROM rides
GROUP BY lvn_state_nm
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[('Г МОСКВА', 82120, Decimal('10893518.50'), 1),
 ('МОСКОВСКАЯ ОБЛ', 35983, Decimal('4778572.75'), 2),
 ('Г САНКТ-ПЕТЕРБУРГ', 34298, Decimal('4393999.37'), 3),
 ('Unknown', 30656, Decimal('4012599.74'), 4),
 ('СВЕРДЛОВСКАЯ ОБЛ', 26128, Decimal('2782558.43'), 5),
 ('КРАСНОДАРСКИЙ КРАЙ', 14612, Decimal('1897869.76'), 6),
 ('РЕСП ТАТАРСТАН', 8703, Decimal('1011694.18'), 7),
 ('РЕСП БАШКОРТОСТАН', 7564, Decimal('948658.76'), 8),
 ('НИЖЕГОРОДСКАЯ ОБЛ', 6777, Decimal('887018.71'), 9),
 ('РЕСП УДМУРТСКАЯ', 4971, Decimal('649280.23'), 10)]

Распределение пользователей по количеству совершённых поездок:

In [18]:
%%sql

SELECT
    CASE
        WHEN count_rides = 1 THEN '1'
        WHEN count_rides BETWEEN 2 AND 5 THEN '2-5'
        WHEN count_rides BETWEEN 6 AND 10 THEN '6-10'
        ELSE '11+'
    END as cohort,
    COUNT(*) as users
FROM

(SELECT
    party_rk,
    COUNT(order_rk) as count_rides
FROM rides
GROUP BY party_rk) t1

GROUP BY cohort

ORDER BY users DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
4 rows affected.


[('1', 23366), ('2-5', 20819), ('11+', 7794), ('6-10', 6054)]

---
## 7. Проверка гипотез
---

Создадим путь для сохранения подготовленных датасетов:

In [19]:
project_root = os.path.abspath('..')
out_dir = os.path.join(project_root, 'data', 'analysis_ready')
os.makedirs(out_dir, exist_ok=True)

H1 — Средняя стоимость поездки различается между мужчинами и женщинами.

In [20]:
query = '''
SELECT
  party_rk,
  gender_cd,
  COUNT(*) as rides_count,
  ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price_per_user
FROM rides
WHERE gender_cd IN ('M','F')
GROUP BY party_rk, gender_cd;
'''

df_h1 = pd.read_sql_query(query, engine)
df_h1.to_parquet('../data/analysis_ready/h1_price_by_gender.parquet', index=False)

H2 — Активность пользователей (количество поездок) зависит от дня недели.

In [21]:
query = '''
SELECT book_start_dttm::DATE as dt, DATE_PART('dow', book_start_dttm::DATE)::INTEGER as dow, COUNT(order_rk) as rides_per_day
FROM rides
GROUP BY dt;
'''

df_h2 = pd.read_sql_query(query, engine, parse_dates=['dt'])
df_h2['dt'] = pd.to_datetime(df_h2['dt'])
df_h2.to_parquet('../data/analysis_ready/h2_daily_counts_by_weekday.parquet', index=False)

H3 — Средняя длительность поездки различается между мужчинами и женщинами.

In [22]:
query = '''
SELECT party_rk, gender_cd,
    COUNT(*) as rides_count,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min
FROM rides
WHERE gender_cd IN ('M', 'F')
GROUP BY party_rk, gender_cd;
'''

df_h3 = pd.read_sql_query(query, engine)
df_h3.to_parquet('../data/analysis_ready/h3_duration_by_gender.parquet', index=False)

H4 — Возраст пользователей влияет на среднюю стоимость поездки.

In [23]:
query = '''
SELECT
  party_rk,
  MAX(age) as age,
  COUNT(*) as rides_count,
  ROUND(AVG(nominal_price_rub_amt)::NUMERIC ,2) as avg_price_per_user,
  ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_per_user
FROM rides
GROUP BY party_rk
HAVING MAX(age) IS NOT NULL;
'''

df_h4 = pd.read_sql_query(query, engine)
df_h4.to_parquet('../data/analysis_ready/h4_age_price_by_user.parquet', index=False)

---
## 8. Выводы
---

- Данные подготовлены и агрегированы в соответствии с целями анализа.
- Все ключевые переменные и показатели (цена, длительность поездки, активность пользователей, демография) доступны в готовых таблицах для проверки гипотез и построения метрик в следующем шаге.
- `4_metrics_and_hypotheses.ipynb` ноутбук может использовать эти таблицы напрямую, без повторной агрегации или очистки.